### PR TITLE
Feature/154 implement prepare handler

### PR DESCRIPTION
**Pull request by rmothilal**
=============================== Coverage summary ===============================
Statements   : 46.31% ( 1112/2401 )
Branches     : 26.89% ( 114/424 )
Functions    : 26.38% ( 143/542 )
Lines        : 47.29% ( 1107/2341 )
================================================================================

1..291
# tests 291
# pass  291

# ok

screen shot 2018-05-28 at 19 23 24

### CODE REVIEW

File: .istanbul.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 instrumentation:
   include-all-sources: true
-  excludes: ['**/migrations/**', '**/deploy/**', '**/src/domain/settlements/settlementEventListener.js']
+  excludes: ['**/migrations/**', '**/deploy/**', '**/src/domain/settlements/settlementEventListener.js', '**/src/handlers/lib/kafka/testProducer.js', '**/testPI2/**']
```
Comment (mdebarros) [inline .istanbul.yml]: testProducer.js should go under the test folder.

File: src/admin/auth/index.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ exports.plugin = {
     // server.auth.default('simple')
     server.auth.strategy('jwt-strategy', 'hapi-now-auth', {
       verifyJWT: true,
-      keychain: [Config.ADMIN_SECRET],
```
Comment (mdebarros) [inline src/admin/auth/index.js]: Why has the config reference been removed and hardcoded to 'secret'?

File: src/admin/participants/routes.js [hunk 3]
```diff
@@ -7,7 +7,33 @@ const RouteConfig = require('../route-config')
 
 const tags = ['api', 'participants']
 const nameValidator = Joi.string().alphanum().min(3).max(30).required().description('Name of the participant')
-const passwordValidator = Joi.string().regex(/^[a-zA-Z0-9]{3,30}$/).required().description('Password for the participant')
+// const passwordValidator = Joi.string().regex(/^[a-zA-Z0-9]{3,30}$/).required().description('Password for the participant')
+const currencyValidator = Joi.string().allow([
```
Comment (mdebarros) [inline src/admin/participants/routes.js]: Shouldn't this be loaded from the DB?

I think we need to add a story for this. Although perhaps its not necessary once we move to the API Design first approach, and use a RAML/Swagger to generate the interfaces classes and validations for us.

Note: The approach can be discussed/determined as part of that. If its the later approach then the story can be closed with the decision documented. I have created the following story for this purpose: https://github.com/mojaloop/project/issues/283  <-- FYI: @elnyry @ksatya77 @NicoDuvenage @wdcmb

File: src/handlers/lib/kafka/testProducer.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
+// @TODO to be cleaned up, used for testing the handlers
```
Comment (mdebarros) [inline src/handlers/lib/kafka/testProducer.js]: Once again, this should be moved to the test folder 💃 

Perhaps under 'test/util' folder?

File: src/handlers/transfers/handler.js [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    throw new Error()
+  }
+  let message = {}
+  try {
+    if (Array.isArray(messages)) {
+      message = messages[0]
+    } else {
+      message = messages
+    }
+    Logger.info('TransferHandler::prepare')
```
Comment (mdebarros) [inline src/handlers/transfers/handler.js]: @wdcmb @NicoDuvenage @ksatya77 @elnyry : See the code below. This is exactly why we need a Rules engine. The section below is going to end up being a tree of entangled hard to read and maintain code. :(

Comment (wdcmb) [reply src/handlers/transfers/handler.js]: Could not agree more.

File: .istanbul.yml (hunk 1 above)
Comment (rmothilal) [reply .istanbul.yml]: changed

File: src/admin/auth/index.js (hunk 2 above)
Comment (rmothilal) [reply src/admin/auth/index.js]: reverted

File: src/handlers/transfers/handler.js (hunk 5 above)
Comment (rmothilal) [reply src/handlers/transfers/handler.js]: so what should i change it to? Any suggestions

File: src/admin/participants/routes.js (hunk 3 above)
Comment (rmothilal) [reply src/admin/participants/routes.js]: no change made here. awaiting currency validation story

File: config/default.json [hunk 6]
```diff
@@ -19,5 +19,291 @@
   },
   "EMAIL_USER": "modusboxemailtest@gmail.com",
   "EMAIL_PASSWORD": "April2o0%",
-  "EMAIL_SMTP": "smtp.gmail.com"
```
Comment (mdebarros) [inline config/default.json]: Note that the "SIDECAR.DISABLED" = **true** must be changed back to **false**

**Comment by mdebarros** [comment]
Note @rmothilal , I am still trying to resolve the connectivity issue to kafka in the Integration Test

File: .istanbul.yml (hunk 1 above)
Comment (mdebarros) [reply .istanbul.yml]: @rajivmothilal I don't see that change reflected here?

File: src/handlers/lib/kafka/testProducer.js (hunk 4 above)
Comment (mdebarros) [reply src/handlers/lib/kafka/testProducer.js]: @rajivmothilal still same dir?

File: config/default.json (hunk 6 above)
Comment (rmothilal) [reply config/default.json]: Change made

Comment (mdebarros) [reply config/default.json]: Still dont see this change?

Comment (rmothilal) [reply config/default.json]: I made the change
its in my branch but it isnt updating the PR

Comment (rmothilal) [reply config/default.json]: ok its there

File: config/default.json [hunk 7]
```diff
@@ -11,13 +11,299 @@
...
+              "messageAsJSON": true,
+              "sync": false,
+              "consumeTimeout": 1000
+            },
+            "rdkafkaConf": {
+              "client.id": "transfer-fulfil",
+              "debug": "all",
+              "group.id": "central-ledger-kafka",
+              "metadata.broker.list": "localhost:9092",
+              "enable.auto.commit": true
```
Comment (mdebarros) [inline config/default.json]: Should be false

File: config/default.json [hunk 8]
```diff
@@ -11,13 +11,299 @@
...
+              "messageAsJSON": true,
+              "sync": false,
+              "consumeTimeout": 1000
+            },
+            "rdkafkaConf": {
+              "client.id": "transfer-reject",
+              "debug": "all",
+              "group.id": "central-ledger-kafka",
+              "metadata.broker.list": "localhost:9092",
+              "enable.auto.commit": true
```
Comment (mdebarros) [inline config/default.json]: Should be false

File: config/default.json [hunk 9]
```diff
@@ -11,13 +11,299 @@
...
+              "messageAsJSON": true,
+              "sync": false,
+              "consumeTimeout": 1000
+            },
+            "rdkafkaConf": {
+              "client.id": "position-prepare",
+              "debug": "all",
+              "group.id": "central-ledger-kafka",
+              "metadata.broker.list": "localhost:9092",
+              "enable.auto.commit": true
```
Comment (mdebarros) [inline config/default.json]: Should be false

File: config/default.json [hunk 10]
```diff
@@ -11,13 +11,299 @@
...
+              "messageAsJSON": true,
+              "sync": false,
+              "consumeTimeout": 1000
+            },
+            "rdkafkaConf": {
+              "client.id": "notification-event",
+              "debug": "all",
+              "group.id": "central-ledger-kafka",
+              "metadata.broker.list": "localhost:9092",
+              "enable.auto.commit": true
```
Comment (mdebarros) [inline config/default.json]: Should be false

File: src/handlers/transfers/handler.js (hunk 5 above)
Comment (mdebarros) [reply src/handlers/transfers/handler.js]: Nothing for now. Will need to refactor this once we have a rules engine.

**Comment by mdebarros** [comment]
screenshot 2018-05-30 09 36 41

Working Unit, & Integration Tests as part of the CI/CD pipeline.

Note that coverage will still fail for the time being.

**Pull request merged**